src/betterproto/plugin/models.py

```diff
@@ -0,0 +1,711 @@
+    FieldDescriptorProto.TYPE_SINT32,  # 17
+    FieldDescriptorProto.TYPE_SINT64,  # 18
+)
+PROTO_BOOL_TYPES = (FieldDescriptorProto.TYPE_BOOL,)  # 8
+PROTO_STR_TYPES = (FieldDescriptorProto.TYPE_STRING,)  # 9
+PROTO_BYTES_TYPES = (FieldDescriptorProto.TYPE_BYTES,)  # 12
+PROTO_MESSAGE_TYPES = (
+    FieldDescriptorProto.TYPE_MESSAGE,  # 11
+    FieldDescriptorProto.TYPE_ENUM,  # 14
+)
+PROTO_MAP_TYPES = (FieldDescriptorProto.TYPE_MESSAGE,)  # 11
+PROTO_PACKED_TYPES = (
+    FieldDescriptorProto.TYPE_DOUBLE,  # 1
+    FieldDescriptorProto.TYPE_FLOAT,  # 2
+    FieldDescriptorProto.TYPE_INT64,  # 3
+    FieldDescriptorProto.TYPE_UINT64,  # 4
+    FieldDescriptorProto.TYPE_INT32,  # 5
+    FieldDescriptorProto.TYPE_FIXED64,  # 6
+    FieldDescriptorProto.TYPE_FIXED32,  # 7
+    FieldDescriptorProto.TYPE_BOOL,  # 8
+    FieldDescriptorProto.TYPE_UINT32,  # 13
+    FieldDescriptorProto.TYPE_SFIXED32,  # 15
+    FieldDescriptorProto.TYPE_SFIXED64,  # 16
+    FieldDescriptorProto.TYPE_SINT32,  # 17
+    FieldDescriptorProto.TYPE_SINT64,  # 18
+)
+
+
+def get_comment(proto_file, path: List[int], indent: int = 4) -> str:
+    pad = " " * indent
+    for sci in proto_file.source_code_info.location:
+        # print(list(sci.path), path, file=sys.stderr)
+        if list(sci.path) == path and sci.leading_comments:
+            lines = textwrap.wrap(
+                sci.leading_comments.strip().replace("\n", ""), width=79 - indent,
+            )
+
+            if path[-2] == 2 and path[-4] != 6:
+                # This is a field
+                return f"{pad}# " + f"\n{pad}# ".join(lines)
+            else:
+                # This is a message, enum, service, or method
+                if len(lines) == 1 and len(lines[0]) < 79 - indent - 6:
+                    lines[0] = lines[0].strip('"')
+                    return f'{pad}"""{lines[0]}"""'
+                else:
+                    joined = f"\n{pad}".join(lines)
+                    return f'{pad}"""\n{pad}{joined}\n{pad}"""'
+
+    return ""
+
+
+class ProtoContentBase:
+    """Methods common to Message, Service and ServiceMethod."""
+
+    path: List[int]
+    comment_indent: int = 4
+
+    def __post_init__(self):
+        """Checks that no fake default fields were left as placeholders."""
+        for field_name, field_val in self.__dataclass_fields__.items():
+            if field_val is PLACEHOLDER:
+                raise ValueError(f"`{field_name}` is a required field.")
+
+    @property
+    def output_file(self) -> "OutputTemplate":
+        current = self
+        while not isinstance(current, OutputTemplate):
+            current = current.parent
+        return current
+
+    @property
+    def proto_file(self) -> FieldDescriptorProto:
+        current = self
+        while not isinstance(current, OutputTemplate):
+            current = current.parent
+        return current.package_proto_obj
+
+    @property
+    def request(self) -> "Request":
```

Comment:
Is this a convenience method or do we need it for compilation?
If possible, I'd prefer to avoid too much treewalking unless really needed, because it makes it harder to think about the code, especially when the tree walking is very abstract.

If we really need the plug-in request object from any object defined in it, we could also just pass it as a constructor arg, and keep things flat. Or would that make things more difficult?

Reply:
At the moment, it's only being used by `ServiceMethod.py_input_message` to get all of the messages. I guess it could be a constructor arg, that might simplify things, or make them more complicated, not sure, I'd have to implement it.